textdb/textdb-dataflow/src/main/java/com/google/re2j/PublicRegexp.java

```diff
@@ -0,0 +1,229 @@
+package com.google.re2j;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+
+/**
+ * Public Wrapper class for re2j.Regexp, this class represents the abstract syntax tree. <br>
+ * 
+ * <p>
+ * For example, <br>
+ * regex: "abc", 
+ * abstract syntax tree:<br>
+ * CONCAT <br>
+ * --LITERAL a <br>
+ * --LITERAL b <br>
+ * --LITERAL c <br>
+ * </p>
+ * 
+ * <p>
+ * regex: "a*|b", 
+ * abstract syntax tree: <br>
+ * ALTERNATE <br>
+ * --STAR <br>
+ * ----LITERAL a <br>
+ * --LITERAL b <br>
+ * </p>
+ * 
+ * <p>
+ * regex: "[a-f]{1-3}", 
+ * abstract syntax tree: <br>
+ * REPEAT min:1, max:3 <br>
+ * --CHAR_CLASS a-f <br>
+ * </p>
+ * 
+ * @author Zuozhi Wang
+ *
+ */
+public class PublicRegexp extends Regexp {
+/*
+	// fields originally declared in Regexp:
+	// for detailed explanations please see getter methods
+	
+	// note that for subexpressions, although original comments
+	// say it's never null, it could still be null
+  
+	Op op;                   // operator
+	int flags;               // bitmap of parse flags
+	Regexp[] subs;           // subexpressions, if any.  Never null.
+		                           // subs[0] is used as the freelist.
+	
+	int[] runes;             // matched runes, for LITERAL, CHAR_CLASS
+	int min, max;            // min, max for REPEAT
+	int cap;                 // capturing index, for CAPTURE
+	String name;             // capturing name, for CAPTURE
+	
+*/
+	
+	// publicSubs are an array of subexpressions with type PublicRegexp
+	PublicRegexp[] publicSubs;
+
+	/**
+	 * This calls the shallow copy constructor in Regexp superclass,
+	 * which only copies reference to subexpressions array. <br>
+	 */
+	private PublicRegexp(Regexp that) {
+		super(that);
+	}
+	  
+	/**
+	 * This performs a deep copy of a Regexp object. Every Regexp Object in subexpression arrary 
+	 * is converted to a PublicRegexp object and put in publicSubs array. <br>
+	 * This is the only public entry point to construct a PublicRegexp object. <br>
+	 * @param re, a Regexp that needs to be converted to PublicRegexp
+	 * @return PublicRegexp
+	 */
+	public static PublicRegexp deepCopy(Regexp re) {
+		PublicRegexp publicRegexp = new PublicRegexp(re);
```

Comment:
Add 
```
if (re.subs == null) 
 return null;
```

Then we don't need the `else` clause.  We can also avoid unnecessary construction of an object that's never used.